File: agenkit-scala/src/main/scala/io/agenkit/patterns/FallbackAgent.scala
package io.agenkit.patterns

import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.AtomicLong

/** Tries each agent in order and returns the first success. */
class FallbackAgent(
  val name: String,
  agents: List[Agent]
) extends Agent:
  private val _processedCount = new AtomicLong(0)

  def capabilities: List[String] = List("fallback", "resilience")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    tryAgents(message, agents)

  private def tryAgents(message: Message, remaining: List[Agent])(using ExecutionContext): Future[Message] =
    remaining match
      case Nil         => Future.failed(new RuntimeException("All agents failed"))
      case head :: tail => head.process(message).recoverWith { case _ => tryAgents(message, tail) }

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("agentCount" -> agents.size, "agents" -> agents.map(_.name)),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/patterns/FallbackAgent.java
package io.agenkit.patterns;

import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * An agent that tries a list of agents in order, falling back on failure.
 */
public final class FallbackAgent implements Agent {

    private final String name;
    private final List<Agent> chain;

    public FallbackAgent(String name, List<Agent> chain) {
        this.name = name;
        this.chain = List.copyOf(chain);
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("fallback", "resilience");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        return tryNext(message, 0, new ArrayList<>());
    }

    private CompletableFuture<Message> tryNext(Message message, int index, List<String> errors) {
        if (index >= chain.size()) {
            String allErrors = String.join("; ", errors);
            return CompletableFuture.completedFuture(
                    Message.of("assistant", "All agents failed: " + allErrors)
                            .withMetadata("fallback_exhausted", true)
                            .withMetadata("errors", errors.size()));
        }

        Agent agent = chain.get(index);
        return agent.process(message)
                .thenApply(response -> response.withMetadata("used_agent", agent.getName()))
                .exceptionallyCompose(ex -> {
                    errors.add(agent.getName() + ": " + ex.getMessage());
                    return tryNext(message, index + 1, errors);
                });
    }

    @Override
    public IntrospectionResult introspect() {
        List<String> agentNames = chain.stream().map(Agent::getName).toList();
        Map<String, Object> state = new HashMap<>();
        state.put("chainLength", chain.size());
        return new IntrospectionResult(name, getCapabilities(), null, state, agentNames);
    }
}
